Articles

AlbertBabayan opened the pull request:
created components for view, create and edit articles, created shared module, respectively shared modal component, and limit pipe.

AlbertBabayan committed:
created components for view, create and edit articles, created shared module, respectively shared modal component, and limit pipe.

AlbertBabayan committed:
small fixes

AlbertBabayan committed:
small fix

sergey-babayan commented on src/app/article/components/article/article.component.ts:
```diff
@@ -1,10 +1,46 @@
+export class ArticleComponent implements OnInit {
 
+  public articles: IArticle[];
+  public article: IArticle;
```
rename to `selectedArticleToView` or something else to be more descriptive

sergey-babayan commented on src/app/article/components/article/article.component.ts:
```diff
@@ -1,10 +1,46 @@
+  constructor() { }
+
+  ngOnInit() {
+    this.articlesSvc.getArticles().pipe(
+      takeUntil(this.destroy$),
+    ).subscribe(
+      resp => {
+        this.articles = resp.result;
+      }
+    )
```
Move this to a method, lifecycle hooks should be clean.

sergey-babayan commented on src/app/article/components/create-article/create-article.component.html:
```diff
@@ -0,0 +1,32 @@
+                        </div>
+                    </div>
+                    <div class="action">
+                        <button class="create" [disabled]="createArticleForm.invalid || !previewImage"
```
`previewImage` should be in the form group as well, the key can be `picture` like the one you send to API,
and it should have a required validator, so the `[disabled]`  would not contain multiple conditions.

sergey-babayan commented on src/app/article/components/article/article.component.ts:
```diff
@@ -1,10 +1,46 @@
+  public articles: IArticle[];
+  public article: IArticle;
+  private articlesSvc = inject(ArticlesService);
+  private destroy$ = new Subject();
```
`destroy$` doesn't do anything, look into how `takeUntil` should be implemented with a `subject`

sergey-babayan commented on src/app/article/components/create-article/create-article.component.ts:
```diff
@@ -0,0 +1,47 @@
+  public currentArticle: IArticle;
+  private articleService = inject(ArticlesService);
+  private formBuilder = inject(FormBuilder);
+  private destroy$ = new Subject();
```
Same here for the `destroy$`

sergey-babayan commented on src/app/article/components/create-article/create-article.component.ts:
```diff
@@ -0,0 +1,47 @@
+
+  public createArticleForm: FormGroup;
+  public previewImage: string;
+  public currentArticle: IArticle;
```
not used

sergey-babayan commented on src/app/article/components/create-article/create-article.component.ts:
```diff
@@ -0,0 +1,47 @@
+  private destroy$ = new Subject();
+
+  ngOnInit() {
+    this.createArticleForm = this.formBuilder.group({
+      title: ['', [Validators.required]],
+      body: ['', [Validators.required]],
+    });
```
Move to a method

sergey-babayan commented on src/app/article/components/edit-article/edit-article.component.ts:
```diff
@@ -0,0 +1,64 @@
+  constructor() { }
+
+  ngOnInit() {
+    this.activeRoute.params.subscribe(params => {
```
If you are using `subscribe` that means that this param can change at any time, even after when the component is fully loaded.
In that case the `getArticle` should've been in the subscription callback function, so when the Id changes you load the new article

To indicate that a subscription should work only once you should do `.pipe(take(1))`
In this case, you can use param snapshot which will give you an object, and no need for a subscription,

But for this kind of components where the route param will not change, and the component needs it to render properly, you should have a resolver, which will get the article for the component before navigation, and the component will use that value.

To go further you can have a guard that will check that the id is present in the URL, otherwise it makes no sense to open the component.

sergey-babayan commented on src/app/article/components/edit-article/edit-article.component.ts:
```diff
@@ -0,0 +1,64 @@
+        this.article = resp
+      }
+    )
+    this.articleForm = this.formBuilder.group({
```
move the rest to dedicated methods.

sergey-babayan commented on src/app/article/guards/activate-article.guard.ts:
```diff
@@ -2,17 +2,16 @@ import { Injectable } from '@angular/core';
-})
+@Injectable()
+
 export class ActivateArticleGuard implements CanActivate {
```
use [functional guards ](https://medium.com/ngconf/functional-route-guards-in-angular-8829f0e4ca5c) instead

sergey-babayan commented on src/app/article/services/articles.service.ts:
```diff
@@ -0,0 +1,32 @@
+  constructor() {}
+
+  public getArticle(payload: number): Observable<any> {
+    return this.http.get<any>(`${this.serverUrl}/api/article/${payload}`);
```
no any.

sergey-babayan commented on src/app/article/components/article/article.component.ts:
```diff
@@ -1,10 +1,46 @@
+  private articlesSvc = inject(ArticlesService);
+  private destroy$ = new Subject();
+  private router = inject(Router);
+  @ViewChild('modal', {static: false}) modalComponent: SharedModalComponent;
```
`@ViewChild` can have the component class as a selector instead of the string with template variable name
`{static: false}` means that this template is inside of some condition, `*ngIf`, `*ngFor` ngSwitch ... and will not be in the dom from the start, no need to mark the template as `{static: false}` if the template is available from the start

```js
@ViewChild(SharedModalComponent') modalComponent: SharedModalComponent;
```

AlbertBabayan committed:
created resolver to get single article, used functional guards, implementation process
component destroying is handled correctly, authservice moved to route module and small fixes.

AlbertBabayan committed:
small fix
